cpp/include/raft/core/detail/copy.hpp

```diff
@@ -0,0 +1,524 @@
+  // perf actually improves once fully implemented.
+  // auto static constexpr const can_use_simd = can_use_host && both_contiguous &&
+  // both_float_or_both_double;
+  auto static constexpr const can_use_simd =
+    can_use_host && both_contiguous && both_float && has_matrix_rank;
+#else
+  auto static constexpr const can_use_simd = false;
+#endif
+
+  auto static constexpr const can_use_std_copy =
+    std::conjunction_v<std::bool_constant<can_use_host>,
+                       std::bool_constant<compatible_dtype>,
+                       std::bool_constant<both_contiguous>,
+                       std::bool_constant<same_underlying_layout>>;
+  auto static constexpr const can_use_raft_copy =
+    std::conjunction_v<std::bool_constant<CUDA_ENABLED>,
+                       std::bool_constant<same_dtype>,
+                       std::bool_constant<both_contiguous>,
+                       std::bool_constant<same_underlying_layout>>;
+
+  auto static constexpr const requires_intermediate =
+    !both_host_accessible && !both_device_accessible && !can_use_raft_copy;
+
+  auto static constexpr const use_intermediate_dst =
+    std::conjunction_v<std::bool_constant<requires_intermediate>,
+                       std::bool_constant<src_device_accessible>>;
+
+  auto static constexpr const use_intermediate_src =
+    std::conjunction_v<std::bool_constant<requires_intermediate>,
+                       std::bool_constant<!use_intermediate_dst>>;
+  auto static constexpr const can_use_device =
+    std::conjunction_v<std::bool_constant<CUDA_ENABLED>,
+                       std::disjunction<std::bool_constant<both_device_accessible>,
+                                        std::bool_constant<requires_intermediate>,
+                                        std::bool_constant<can_use_raft_copy>>>;
+
+  auto static constexpr const can_use_cublas =
+    std::conjunction_v<std::bool_constant<can_use_device>,
+                       std::bool_constant<compatible_dtype>,
+                       std::bool_constant<both_contiguous>,
+                       std::bool_constant<!same_underlying_layout>,
+                       std::bool_constant<has_matrix_rank>,
+                       std::bool_constant<both_float_or_both_double>>;
+
+  auto static constexpr const custom_kernel_allowed =
+    std::conjunction_v<std::bool_constant<can_use_device>,
+                       std::bool_constant<!(can_use_raft_copy || can_use_cublas)>>;
+
+  auto static constexpr const custom_kernel_required =
+    std::conjunction_v<std::bool_constant<!can_use_host>,
+                       std::bool_constant<!(can_use_raft_copy || can_use_cublas)>>;
+
+  // Viable overload?
+  auto static constexpr const value =
+    std::conjunction_v<std::bool_constant<is_mdspan_v<src_type>>,
+                       std::bool_constant<is_mdspan_v<dst_type>>,
+                       std::bool_constant<can_use_host || can_use_device>>;
+  using type = std::enable_if_t<value, T>;
```

Comment:
Is there a scenario where this is not `void`?

Reply:
There will be with mdbuffer. In general, this is a convenience just to parallel the behavior of `enable_if_t`, but just like the extra type argument is rarely used with `enable_if_t`, I expect it to be rarely used here as well.